### ISSUE TITLE
ttnn.reciprocal pcc error

### ISSUE THREAD

**Issue by jserbedzijaTT** [tenstorrent/tt-metal#25839]
### Describe the bug

I get a PCC error on ttnn.reciprocal in this test:

```
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
# SPDX-License-Identifier: Apache-2.0

import pytest
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

def test_reciprocal(device):
    input_shape = (784, 784)
    torch_input_a = torch.rand(input_shape, dtype=torch.float32)
    torch_input_b = torch.rand(input_shape, dtype=torch.float32)
    torch_input_c = torch.rand(input_shape, dtype=torch.float32)

    torch_multiply_result = torch_input_c * torch_input_a
    torch_subtract_result = torch_input_b - torch_multiply_result
    torch_reciprocal_result = torch.reciprocal(torch_subtract_result)

    ttnn_tensor_a = ttnn.from_torch(
        torch_input_a,
        layout=ttnn.ROW_MAJOR_LAYOUT,
        device=device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    ttnn_tensor_b = ttnn.from_torch(
        torch_input_b,
        layout=ttnn.ROW_MAJOR_LAYOUT,
        device=device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    ttnn_tensor_c = ttnn.from_torch(
        torch_input_c,
        layout=ttnn.ROW_MAJOR_LAYOUT,
        device=device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG
    )

    ttnn_multiply_result = ttnn.multiply(ttnn_tensor_c, ttnn_tensor_a)
    assert_with_pcc(torch_multiply_result, ttnn.to_torch(ttnn_multiply_result), 0.999)

    ttnn_subtract_result = ttnn.subtract(ttnn_tensor_b, ttnn_multiply_result)
    assert_with_pcc(torch_subtract_result, ttnn.to_torch(ttnn_subtract_result), 0.999)

    ttnn_subtract_host = ttnn.from_device(ttnn_subtract_result)
    ttnn_subtract_tile = ttnn.to_layout(ttnn_subtract_host, ttnn.TILE_LAYOUT)
    ttnn_subtract_device = ttnn.to_device(
        ttnn_subtract_tile,
        device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    assert_with_pcc(torch_subtract_result, ttnn.to_torch(ttnn_subtract_device), 0.999)

    ttnn_reciprocal_result = ttnn.reciprocal(ttnn_subtract_device)
    assert_with_pcc(torch_reciprocal_result, ttnn.to_torch(ttnn_reciprocal_result), 0.999)

### Steps to reproduce the issue

Run the test left in `Describe the bug`


### Expected behavior

The test should pass

### Please complete the following environment information

_No response_

**Comment by jserbedzijaTT** [comment]
Can we prioritize this one? @cmaryanTT

**Comment by jasondavies** [comment]
@jserbedzijaTT I looked into this briefly (I have a pending PR with high-precision reciprocal) and the issue was poor precision in `ttnn_subtract_result`, not reciprocal.

Taking a reciprocal then inflates any small errors, leading to the low PCC error.  Try using `torch.allclose` or similar instead of `assert_with_pcc` to check `ttnn_subtract_result`.

As for the root cause, I think it's due to the conversion between ROW_MAJOR_LAYOUT and TILE_LAYOUT.  If you change everything to TILE_LAYOUT, the test seems to pass fine.  There may be a bug in `ttnn.to_layout` relating to precision loss when converting.

**Comment by jserbedzijaTT** [comment]
Oh okay, thanks for the clarification. Can someone take a look at why `ttnn.to_layout` is producing a weird result? @cmaryanTT

**Comment by jasondavies** [comment]
@jserbedzijaTT Further exploration and I think the issue is likely due to `ttnn.multiply` with ROW_MAJOR_LAYOUT, not `ttnn.to_layout` per se.  It seems to have worse precision compared to using TILE_LAYOUT.

This is likely due to the fact that internally it uses the tilize LLK (followed by eltwise multiply, then untilize), which truncates float32 via SrcA, which can only hold tf32.  Internally, it looks like binary_ng uses tilize/untilize if the input is RM.

It's relatively straightforward to make tilize preserve float32 via unpack_to_dest.

Alternatively, ensure that TILE_LAYOUT is used to avoid precision loss as a workaround.

**Comment by jserbedzijaTT** [comment]
Is it possible for binary ops to operate on a row-major layout instead of tilizing it? Currently, we have some examples where the last two dimensions of a tensor are (1, 1), and if we change its layout to tile they are going to grow to (32, 32) increasing the memory usage significantly.

**Comment by jasondavies** [comment]
This is where the conversion to TILE happens: https://github.com/tenstorrent/tt-metal/blob/f3d22baacac0fc96fc302d20ed68c00206c4e01d/ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp#L443-L474

@patrickroberts might know.

**Comment by jasondavies** [comment]
I've opened this to track the float32 tilize issue: https://github.com/tenstorrent/tt-metal/issues/26062

**Comment by jserbedzijaTT** [comment]
Any thoughts? @patrickroberts

**Comment by cmaryanTT** [comment]
Tilize for fp32 should be lossless, I believe there is another issue open for that - @ntarafdar ’s team owns that.

We can look at RM native eltwise OPs, but it’s not a high priority.

**Comment by minaliuTT** [comment]
@fahmedchanTT Is this ticket still relevant? If not, can we close it? Thanks!

**Comment by cmaryanTT** [comment]
Remaining issue is accuracy of fp32 tilize, tracked in https://github.com/tenstorrent/tt-metal/issues/23405  - closing this issue as the problem is generic, and recip itself is otherwise fine.
